File: Redcoder/Python/83.py
N, M = map(int,input().split())
num = [int(i) for i in input().split()]
box = [0] * (N+1)
for i in range(1,M):
    box[min(num[i], num[i-1])-1] += 1
    box[max(num[i], num[i-1])-1] -= 1
s = [0]
for i in range(N+1):
    s.append(s[-1] + box[i])
cost = [[0,0,0]]
for i in range(N-1):
    a,b,c  = map(int,input().split())
    cost.append([a,b,c])
ans = 0

for i in range(1,N):
    ans += min(cost[i][0] * s[i], cost[i][2] + cost[i][1] * s[i])
print(ans)

File: Redcoder/C++/83.cpp
#include <bits/stdc++.h>
#define rep(i, nn, n) for (int i = nn; i < (n); ++i)
#define rrep(i, nn, n) for (int i = nn; i > (n); --i)
using namespace std;
using ll = long long;
using P = pair<int, int>;
#define chmin(x, y) x = min(x, y)
#define chmax(x, y) x = max(x, y)

int main()
{
    int N, M;
    cin >> N >> M;
    vector<int> P(M, 0);
    rep(i, 0, M)
    {
        cin >> P[i];
        --P[i];
    }
    vector<int> box(N, 0);
    int now = P[0];
    rep(i, 1, M)
    {
        int a, b;
        a = min(now, P[i]);
        b = max(now, P[i]);
        --box[b];
        ++box[a];
        now = P[i];
    }
    vector<int> s(N + 1, 0);
    rep(i, 0, N)
    {
        s[i + 1] = s[i] + box[i];
    }
    ll ans = 0;
    rep(i, 0, N - 1)
    {
        int a, b, c;
        cin >> a >> b >> c;
        ans += min(1LL * a * s[i + 1], 1LL * b * s[i + 1] + c);
    }

    cout << ans << endl;
    return 0;
}